File: iOS/Pool/Sources/Pool/PoolScore/PoolScoreItem.swift
import SwiftUI
import UI

struct PoolScoreItem: View {
    let poolGamblerScore: PoolGamblerScoreModel
    let onOpen: () -> Void
    let onJoin: () -> Void
    let placeholderModifier: (any ViewModifier)?

    @Environment(\.boxSpacing) private var boxSpacing
    @Environment(\.dynamicTypeSize) private var dynamicTypeSize

    /// The rank tile and invite icon grow with Dynamic Type. Unlike the leaderboard's rank
    /// rail, this row's rail has no fixed width, so the tile can afford to scale.
    @ScaledMetric(relativeTo: .body) private var rankTileSize: CGFloat = 32
    @ScaledMetric(relativeTo: .body) private var inviteIconSize: CGFloat = 20

    init(
        poolGamblerScore: PoolGamblerScoreModel,
        onOpen: @escaping () -> Void,
        onJoin: @escaping () -> Void,
        placeholderModifier: (any ViewModifier)? = nil
    ) {
        self.poolGamblerScore = poolGamblerScore
        self.onOpen = onOpen
        self.onJoin = onJoin
        self.placeholderModifier = placeholderModifier
    }

    var isPlaceholder: Bool {
        placeholderModifier != nil
    }

    var body: some View {
        Group {
            if let placeholderModifier {
                scoreContent(applying: placeholderModifier)
            } else {
                scoreContent
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityAddTraits(.isButton)
        .accessibilityHint(Text(.poolScoreOpenAccessibilityAction))
        .accessibilityAction {
            onOpen()
        }
        .accessibilityAction(named: Text(.poolScoreInviteAccessibilityAction)) {
            onJoin()
        }
        .accessibilityHidden(isPlaceholder)
    }

    private func scoreContent(applying modifier: some ViewModifier) -> AnyView {
        AnyView(scoreContent.modifier(modifier))
    }

    private var scoreContent: some View {
        Group {
            if dynamicTypeSize.isAccessibilitySize {
                reflowedContent
            } else {
                inlineContent
            }
        }
        .padding(boxSpacing.large)
        .frame(maxWidth: .infinity)
    }

    /// The normal row: rank rail, identity, then the invite control on the trailing edge.
    private var inlineContent: some View {
        HStack(spacing: boxSpacing.medium) {
            rankRail

            VStack(alignment: .leading) {
                Text(poolGamblerScore.poolName)

                metaCounts
            }

            Spacer()

            inviteButton
        }
    }

    /// At accessibility text sizes three columns cannot share 400 points: the identity column
    /// got so narrow that the pool name and member count broke mid-word ("Mund / ial",
    /// "me / mbe / rs"). The row reflows to full-width rows instead, so every string gets the
    /// whole width and wraps on word boundaries.
    private var reflowedContent: some View {
        VStack(alignment: .leading, spacing: boxSpacing.medium) {
            HStack(spacing: boxSpacing.medium) {
                rankRail

                Spacer()

                inviteButton
            }

            Text(poolGamblerScore.poolName)

            metaCounts
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var rankRail: some View {
        VStack(spacing: boxSpacing.small) {
            if let position = poolGamblerScore.position {
                PostionIndicator(
                    position: position,
                    shouldUsePrimeryColor: false,
                    size: rankTileSize
                )
            }

            if let rank = poolGamblerScore.rank() {
                TrendIndicator(difference: rank, textStyle: Font.TextStyle.footnote)
            }
        }
    }

    private var inviteButton: some View {
        Button(action: onJoin) {
            Image(.personAdd)
                .resizable()
                .scaledToFit()
                .frame(width: inviteIconSize, height: inviteIconSize)
        }
        .buttonStyle(.liquidGlass)
        .buttonBorderShape(.capsule)
        .accessibilityLabel(Text(.poolScoreInviteAccessibilityAction))
    }

    /// Points and member count. They sit side by side normally, but at accessibility text
    /// sizes there is no room for two columns — squeezed into one they wrapped mid-word
    /// ("me / mbe / rs"), so they stack instead and each keeps a single, shrink-to-fit line.
    @ViewBuilder
    private var metaCounts: some View {
        let score = poolGamblerScore.score.map { Text(.pointsText($0)).font(.footnote.bold()) }
        let members = poolGamblerScore.gamblerCount.map { Text(.gamblersText($0)).font(.footnote) }

        if dynamicTypeSize.isAccessibilitySize {
            VStack(alignment: .leading, spacing: boxSpacing.small) {
                score
                members
            }
        } else {
            HStack(spacing: boxSpacing.large) {
                score
                members
            }
            .lineLimit(1)
            .minimumScaleFactor(countMinimumScaleFactor)
        }
    }

    /// The exact string handed to `.accessibilityLabel` in `body`. A placeholder row
    /// contributes no VoiceOver text, so it resolves to the empty string.
    ///
    /// Unlike the leaderboard row, this one leads with the pool name: these rows are
    /// not a ranking, so the name is what distinguishes one row from the next.
    var accessibilityLabel: String {
        isPlaceholder ? "" : loadedAccessibilityLabel
    }

    private var loadedAccessibilityLabel: String {
        [
            poolGamblerScore.poolName,
            rankAccessibilityText,
            scoreAccessibilityText,
            gamblerCountAccessibilityText,
            movementAccessibilityText,
        ]
        .compactMap { $0 }
        .joined(separator: ", ")
    }

    private var rankAccessibilityText: String {
        guard let position = poolGamblerScore.position else {
            return String(localized: .leaderboardRankMissingAccessibility)
        }
        return String(localized: .leaderboardRankAccessibility(position))
    }

    private var scoreAccessibilityText: String {
        guard let score = poolGamblerScore.score else {
            return String(localized: .leaderboardScoreMissingAccessibility)
        }
        return String(localized: .leaderboardPointsAccessibility(score))
    }

    /// The visible member count already spells the word out, so the announcement
    /// reuses its string rather than duplicating it under an accessibility key —
    /// including its plural agreement, since the row below renders the same symbol.
    private var gamblerCountAccessibilityText: String? {
        guard let gamblerCount = poolGamblerScore.gamblerCount else { return nil }
        return String(localized: .gamblersText(gamblerCount))
    }

    private var movementAccessibilityText: String? {
        guard let difference = poolGamblerScore.rank() else { return nil }

        switch difference {
        case let value where value > 0:
            return String(localized: .leaderboardMovementUpAccessibility(abs(value)))
        case let value where value < 0:
            return String(localized: .leaderboardMovementDownAccessibility(abs(value)))
        default:
            return String(localized: .leaderboardMovementUnchangedAccessibility)
        }
    }
}

private let countMinimumScaleFactor: CGFloat = 0.7

#Preview("Normal Light") {
    PoolScoreItem(poolGamblerScore: poolGamblerScoreDummyModel(), onOpen: {}, onJoin: {})
        .preferredColorScheme(.light)
}

#Preview("Normal Dark") {
    PoolScoreItem(poolGamblerScore: poolGamblerScoreDummyModel(), onOpen: {}, onJoin: {})
        .preferredColorScheme(.dark)
}

#Preview("Without position") {
    PoolScoreItem(
        poolGamblerScore: poolGamblerScoreWithoutPositionDummyModel(),
        onOpen: {},
        onJoin: {}
    )
}

#Preview("Long pool name") {
    PoolScoreItem(
        poolGamblerScore: PoolGamblerScoreModel(
            poolId: "A3C2E1",
            poolName: "This is a very long pool name to test how it is displayed in the list",
            gamblerId: "YF23H1",
            gamblerUsername: "neptune-player",
            position: 4,
            beforePosition: 3,
            score: 8,
            gamblerCount: 101
        ),
        onOpen: {},
        onJoin: {}
    )
}

#Preview("Placeholder") {
    PoolScoreItem(
        poolGamblerScore: poolGamblerScorePlaceholderModel(),
        onOpen: {},
        onJoin: {},
        placeholderModifier: ShimmerModifier()
    )
}

File: Android/pool/src/main/java/com/felipearpa/tyche/pool/poolscore/PoolScoreItem.kt
package com.felipearpa.tyche.pool.poolscore

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.PreviewLightDark
import com.felipearpa.tyche.pool.PoolGamblerScoreModel
import com.felipearpa.tyche.pool.PositionIndicator
import com.felipearpa.tyche.pool.R
import com.felipearpa.tyche.pool.poolGamblerScoreDummyModel
import com.felipearpa.tyche.pool.poolGamblerScorePlaceholderModel
import com.felipearpa.tyche.pool.poolGamblerScoreWithoutPositionDummyModel
import com.felipearpa.tyche.pool.rank
import com.felipearpa.tyche.ui.TrendIndicator
import com.felipearpa.tyche.ui.shimmer
import com.felipearpa.tyche.ui.theme.LocalBoxSpacing
import com.felipearpa.tyche.ui.theme.TycheTheme
import kotlin.math.abs
import com.felipearpa.tyche.ui.R as SharedR

@Composable
fun PoolScoreItem(
    poolGamblerScore: PoolGamblerScoreModel,
    onJoin: () -> Unit,
    modifier: Modifier = Modifier,
    placeholderModifier: Modifier? = null,
) {
    val isPlaceholder = placeholderModifier != null
    val placeholderStyle = placeholderModifier ?: Modifier
    val accessibilityDescription = poolScoreAccessibilityDescription(poolGamblerScore)

    Column(
        verticalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium),
        modifier = modifier,
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(all = LocalBoxSpacing.current.medium)
                // The row owns its announcement, so the rank tile's bare "4" and the
                // trend indicator's bare "1" never reach TalkBack on their own. Clearing
                // also drops the invite button's node; `PoolScoreList` re-exposes invite
                // as a custom action on the clickable row that wraps this one.
                .clearAndSetSemantics {
                    if (!isPlaceholder) contentDescription = accessibilityDescription
                },
            horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.small),
            ) {
                poolGamblerScore.position?.let {
                    PositionIndicator(
                        position = it,
                        shouldUsePrimaryColor = false,
                        placeholderModifier = placeholderStyle,
                    )
                }
                poolGamblerScore.rank()?.let {
                    TrendIndicator(
                        placeholderModifier = placeholderStyle,
                        rank = it,
                        textStyle = MaterialTheme.typography.labelSmall,
                    )
                }
            }

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = poolGamblerScore.poolName,
                    modifier = placeholderStyle,
                )

                Row(
                    horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.large),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    poolGamblerScore.score?.let {
                        Text(
                            text = stringResource(id = R.string.points_text, it),
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.Bold,
                            modifier = placeholderStyle,
                        )
                    }

                    poolGamblerScore.gamblerCount?.let {
                        Text(
                            text = pluralStringResource(R.plurals.gamblers_text, it, it),
                            style = MaterialTheme.typography.labelSmall,
                            modifier = placeholderStyle,
                        )
                    }
                }
            }

            IconButton(
                onClick = onJoin,
                modifier = placeholderStyle,
            ) {
                Icon(
                    painter = painterResource(SharedR.drawable.person_add),
                    contentDescription = null,
                )
            }
        }
    }
}

@Composable
fun PoolScorePlaceholderItem(modifier: Modifier = Modifier) {
    PoolScoreItem(
        poolGamblerScore = poolGamblerScorePlaceholderModel(),
        onJoin = {},
        modifier = modifier,
        placeholderModifier = Modifier.shimmer(),
    )
}

/**
 * The row's single announcement, composed from the model rather than from the abbreviated
 * visible copy. Unlike the leaderboard row it leads with the pool name: these rows are not
 * a ranking, so the name is what distinguishes one row from the next.
 */
@Composable
private fun poolScoreAccessibilityDescription(
    poolGamblerScore: PoolGamblerScoreModel,
): String {
    val rank = poolGamblerScore.position?.let {
        stringResource(R.string.leaderboard_rank_accessibility, it)
    } ?: stringResource(R.string.leaderboard_rank_missing_accessibility)
    val score = poolGamblerScore.score?.let {
        pluralStringResource(R.plurals.leaderboard_points_accessibility, it, it)
    } ?: stringResource(R.string.leaderboard_score_missing_accessibility)
    // The visible member count already spells the word out, so the announcement reuses
    // its resource rather than duplicating it under an accessibility key — including its
    // plural agreement, since the row above renders the same one.
    val gamblerCount = poolGamblerScore.gamblerCount?.let {
        pluralStringResource(R.plurals.gamblers_text, it, it)
    }
    val movement = poolGamblerScore.rank()?.let { difference ->
        when {
            difference > 0 -> pluralStringResource(
                R.plurals.leaderboard_movement_up_accessibility,
                abs(difference),
                abs(difference),
            )

            difference < 0 -> pluralStringResource(
                R.plurals.leaderboard_movement_down_accessibility,
                abs(difference),
                abs(difference),
            )

            else -> stringResource(R.string.leaderboard_movement_unchanged_accessibility)
        }
    }

    return listOfNotNull(
        poolGamblerScore.poolName,
        rank,
        score,
        gamblerCount,
        movement,
    ).joinToString(separator = ", ")
}

@PreviewLightDark
@Composable
private fun PoolScoreItemPreview() {
    TycheTheme {
        Surface {
            PoolScoreItem(
                poolGamblerScore = poolGamblerScoreDummyModel(),
                onJoin = {},
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun PoolScoreItemPreviewWithoutScore() {
    TycheTheme {
        Surface {
            PoolScoreItem(
                poolGamblerScore = poolGamblerScoreWithoutPositionDummyModel(),
                onJoin = {},
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun PoolScoreItemLongPoolNamePreview() {
    TycheTheme {
        Surface {
            PoolScoreItem(
                poolGamblerScore = poolGamblerScoreDummyModel().copy(
                    poolName = "This is a very long pool name to test how it is displayed in the list",
                ),
                onJoin = {},
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun PoolScoreFakeItemPreview() {
    TycheTheme {
        Surface {
            PoolScorePlaceholderItem(modifier = Modifier.fillMaxWidth())
        }
    }
}
